keps/sig-auth/3766-referential-authorization/README.md

```diff
@@ -82,59 +94,88 @@ been [used by
-cases other than Gateway API. We would like to move ReferenceGrant to a neutral home
-(ideally, under SIG Auth) in order to make it the canonical API for managing references
-across namespaces.
+Any project that wants to enable cross-namespace references currently has to
+choose between introducing a dependency on Gateway API's ReferenceGrant or
+creating a new API that would be partially redundant (leading to confusion for
+users).
+
+Recent interest between SIGs has made it clear that ReferenceGrant is wanted for
+use cases other than Gateway API. We would like to move ReferenceGrant to a
+neutral home (ideally, under SIG Auth) in order to make it the canonical API for
+managing references across namespaces.
+
+At the same time, Ingress and Gateway controllers are often deployed with read
+access to all Secrets in a cluster. This can be a significant security risk as
+these components often already sit on the edge of a cluster. Finding a way to
+compromise an Ingress or Gateway controller could potentially provide access to
+all Secrets in a cluster. Ideally, a solution that is related to referential
+authorization could also help mitigate this problem.
 
 ### Goals
 
-* Move ReferenceGrant to an API Group that SIG Auth manages
+* Introduce API(s) that enable the same capabilities that ReferenceGrant does
+  within Gateway API.
+* In addition to ReferenceGrant's capabilities, expand the scope of this KEP so
+  that it enables:
+  * Authorizing controllers to read only resources that have been referenced by
+    resources they are implementing.
+  * Building a more generic approach that will enable generic controllers and
+    authorizers to implement this API without prior knowledge of resources that
+    references are from or to.
 * Clearly define how ReferenceGrant should be used, including both current use
-  cases and guidance for future use cases
+  cases and guidance for future use cases.
 * Implement a library to ensure that ReferenceGrant is implemented consistently
-  by all controllers
+  by all controllers.
 
 ### Non-Goals
 
-* Develop an authorizer that will automatically implement ReferenceGrant for all
-  use cases. (It would be impossible to represent concepts like "all namespaces"
-  or label selectors that have become important for this KEP).
+* Advanced capabilities like label selectors and the ability to allow references
+  from any namespace are likely reasonable future extensions of this API but are
+  intentionally out of scope for the initial work.
 
 ## Proposal
 
-Move the existing ReferenceGrant resource into a new
-`authorization.k8s.io` API group, defined within the Kubernetes code base
-as part of the 1.27 release.
-
-We will take this opportunity to clarify and update the API after SIG Auth
-feedback. This resource will start with v1alpha1 as the API version.
-
+Create 3 new API types in a new `reference.authorization.k8s.io` API Group:
+
+1. **ClusterReferenceGrant:** This resource grants access to resources by
```

Comment:
Naming-wise, if this is "same namespace," is it confusing to call it "Cluster"ReferenceGrant, given that ClusterRole/ClusterRoleBinding are cluster-wide instead of same namespace?

Reply:
Though I guess the behavior here is a single subject is still granted permission on something across multiple namespaces, so perhaps that makes sense. The "same-namespace" bit is just referring to the references between namespaced resources, not access paths from subjects to namespaces.

Reply:
Yeah it's strange to find the right name here, I was trying to go with the idea that this resource is very similar to ReferenceGrant but is cluster-scoped. Would be very open to a different name here, it could potentially make it easier to understand. Anyone have ideas here? We can also revert to ClusterReferencePattern which may be more clear.

cc @luxas @stealthybox @ttakahashi21

Reply:
Yes, this is something that confused me as well, I think we somehow have to rename it.

WithinNamespaceReferenceGrant is a mouthful, but kind of describes it well
SameNamespaceReferenceGrant
NamespacedReferenceGrant is confusing, as ReferenceGrant is a namespaced resource already :D 

somehow I'd like to get in the "namespace" in this name, although that makes the resource name quite long.
but important would be to emphasise that ReferenceGrant is fine-grained, SameNamespaceReferenceGrant (or similar) is a little more coarse, or more like a "generic rule"

Reply:
I agree with the name change. Certainly, I think the name should convey the meaning that it is for within the same namespace.
How about @stealthybox ?